// promptfoo/dist/src/util/server.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserBehavior = void 0;
exports.checkServerRunning = checkServerRunning;
exports.openBrowser = openBrowser;
const opener_1 = __importDefault(require("opener"));
const constants_1 = require("../constants");
const logger_1 = __importDefault(require("../logger"));
const readline_1 = require("./readline");
var BrowserBehavior;
(function (BrowserBehavior) {
    BrowserBehavior[BrowserBehavior["ASK"] = 0] = "ASK";
    BrowserBehavior[BrowserBehavior["OPEN"] = 1] = "OPEN";
    BrowserBehavior[BrowserBehavior["SKIP"] = 2] = "SKIP";
    BrowserBehavior[BrowserBehavior["OPEN_TO_REPORT"] = 3] = "OPEN_TO_REPORT";
    BrowserBehavior[BrowserBehavior["OPEN_TO_REDTEAM_CREATE"] = 4] = "OPEN_TO_REDTEAM_CREATE";
})(BrowserBehavior || (exports.BrowserBehavior = BrowserBehavior = {}));
async function checkServerRunning(port = (0, constants_1.getDefaultPort)()) {
    try {
        const response = await fetch(`http://localhost:${port}/health`);
        const data = await response.json();
        return data.status === 'OK' && data.version === constants_1.VERSION;
    }
    catch (err) {
        logger_1.default.debug(`Failed to check server health: ${String(err)}`);
        return false;
    }
}
async function openBrowser(browserBehavior, port = (0, constants_1.getDefaultPort)()) {
    const baseUrl = `http://localhost:${port}`;
    let url = baseUrl;
    if (browserBehavior === BrowserBehavior.OPEN_TO_REPORT) {
        url = `${baseUrl}/report`;
    }
    else if (browserBehavior === BrowserBehavior.OPEN_TO_REDTEAM_CREATE) {
        url = `${baseUrl}/redteam/setup`;
    }
    const doOpen = async () => {
        try {
            logger_1.default.info('Press Ctrl+C to stop the server');
            await (0, opener_1.default)(url);
        }
        catch (err) {
            logger_1.default.error(`Failed to open browser: ${String(err)}`);
        }
    };
    if (browserBehavior === BrowserBehavior.ASK) {
        const shouldOpen = await (0, readline_1.promptYesNo)('Open URL in browser?', false);
        if (shouldOpen) {
            await doOpen();
        }
    }
    else if (browserBehavior !== BrowserBehavior.SKIP) {
        await doOpen();
    }
}
